text/separate-compilation.md

```diff
@@ -0,0 +1,229 @@
+distinguishes three scenarios, depending on the existence of:
+- both `lib.vi` and `lib.v`;
+- only `lib.vi`;
+- or only `lib.v`.
+
+## Both a `.vi` and `.v` File
+
+For concreteness, consider the following example:
+
+```coq
+(* lib.vi *)
+Global Open Scope Z_scope.
+Parameter answer : nat.
+Axiom answer_is_42 : answer = 42%nat.
+
+(* lib.v *)
+Require Import stdpp.prelude.
+Global Open Scope N_scope.
+
+Definition answer : nat := 42.
+Definition answer_is_42 : answer = 42%nat := ltac:(reflexivity).
+
+Definition other_answer : N := 42 + 1.
+```
+
+Here, `lib.v` contains both the implementation of `lib.vi` and some commands
+that perform global side effects and that should be hidden from clients:
+`stdpp.prelude` modifies many Coq settings aggressively, even when simply
+`Require`d, so adding this `Require` is a significant breaking change. 
+
+In this example, the semantics of `lib` would resemble the semantics of the
+following Coq source:
+
+```coq
+(* lib_composed.v: *)
+Module Type __LIB.
+  Global Open Scope Z_scope.
+  Parameter answer : nat.
+  Axiom answer_is_42 : answer = 42.
+End __LIB.
+
+Module __lib : __LIB.
+  Require Import stdpp.prelude.
+  (* ^ Require inside a Module is strongly discouraged. What is relevant is
+     that side effects of the Module are not visible outside of this Module
+     *even* when the Module is Import-ed *)
+  Global Open Scope N_scope.
+
+  Definition answer : nat := 42.
+  Definition answer_is_42 : answer = 42%nat := ltac:(reflexivity).
+End __lib.
+
+Export __lib. (* make the declarations from lib available from [Import]ing lib (rather that lib.lib *)
+```
+
+In this example translation, we use the `\_\_` prefix for generated identifiers
+used as a translation device; these identifiers should be hidden from clients.
+
+This translation turns the `.vi` interface into a module type, the `.v`
+implementation into a module, and uses opaque module ascription to hide the
+module contents.
+The only visible side effects are those that appear in the interface: here,
+only `Global Open Scope Z_scope`. As we only `Require` stdpp.prelude in the
+module body, we intend this to be **hidden** from clients that perform `Require
+lib`, even if this might not be guaranteed by `lib_composed.v`.
+
+## Only a `.vi` File
+
+When only an interface file exists, there is (potentially) no underlying
+implementation, but the existance of the interface should still provide definite
+references to an implementation. Casting this in the feature set of the current
+Coq ecosystem, this essentially translates to a `Declare Module`
+[^full-functorization]. Concretely,
+
+```coq
+(* lib.vi *)
+Global Open Scope Z_scope.
+Parameter answer : nat.
+Axiom answer_is_42 : answer = 42.
+```
+
+would be "translated" to:
+
+```coq
+(* lib.v *)
+Module Type __LIB.
+  Global Open Scope Z_scope.
+  Parameter answer : nat.
+  Axiom answer_is_42 : answer = 42.
+End __LIB.
+
+Declare Module __lib : __LIB.
+Export __lib.
+```
+
+## Only a `.v` File
```

Comment:
Do you expect that this will be the most common case for the long term?

Reply:
No. I would hope that there are a lot of libraries that would have meaningful theory that could be put in a `.vi` file. @charguer mentioned some things like this in his standard library talk.